File: apps/engine-api/src/api/routes/search_profile.rs
use axum::extract::State;

use crate::api::dto::profile::AnalyzeProfileResponse;
use crate::api::dto::search_profile::{
    BuildSearchProfileRequest, BuildSearchProfileResponse, SearchProfileResponse,
};
use crate::api::error::{ApiError, ApiJson};
use crate::state::AppState;

pub async fn build_search_profile(
    State(state): State<AppState>,
    ApiJson(payload): ApiJson<BuildSearchProfileRequest>,
) -> Result<axum::Json<BuildSearchProfileResponse>, ApiError> {
    let input = payload.validate()?;
    let analyzed_profile = state.profile_analysis_service.analyze(&input.raw_text);
    let search_profile = state
        .search_profile_service
        .build(&analyzed_profile, &input.preferences);

    Ok(axum::Json(BuildSearchProfileResponse {
        analyzed_profile: AnalyzeProfileResponse::from(analyzed_profile),
        search_profile: SearchProfileResponse::from(search_profile),
    }))
}

#[cfg(test)]
mod tests {
    use axum::body;
    use axum::extract::State;
    use axum::http::StatusCode;
    use axum::response::IntoResponse;
    use serde_json::{Value, json};

    use crate::api::dto::search_profile::{BuildSearchProfileRequest, SearchPreferencesRequest};
    use crate::api::error::ApiJson;
    use crate::services::applications::{ApplicationsService, ApplicationsServiceStub};
    use crate::services::jobs::{JobsService, JobsServiceStub};
    use crate::services::profiles::{ProfilesService, ProfilesServiceStub};
    use crate::services::resumes::{ResumesService, ResumesServiceStub};
    use crate::state::AppState;

    use super::build_search_profile;

    fn test_state() -> AppState {
        AppState::for_services(
            ProfilesService::for_tests(ProfilesServiceStub::default()),
            JobsService::for_tests(JobsServiceStub::default()),
            ApplicationsService::for_tests(ApplicationsServiceStub::default()),
            ResumesService::for_tests(ResumesServiceStub::default()),
        )
    }

    #[tokio::test]
    async fn builds_search_profile_from_raw_text() {
        let response = build_search_profile(
            State(test_state()),
            ApiJson(BuildSearchProfileRequest {
                raw_text: "Senior React Native developer with product experience".to_string(),
                preferences: SearchPreferencesRequest {
                    target_regions: vec![crate::domain::search::profile::TargetRegion::Ua],
                    work_modes: vec![crate::domain::search::profile::WorkMode::Remote],
                    preferred_roles: vec!["frontend_engineer".to_string()],
                    allowed_sources: vec!["djinni".to_string(), "work_ua".to_string()],
                    include_keywords: vec!["product company".to_string()],
                    exclude_keywords: vec!["gambling".to_string()],
                },
            }),
        )
        .await
        .expect("handler should build search profile")
        .into_response();

        assert_eq!(response.status(), StatusCode::OK);

        let body = body::to_bytes(response.into_body(), usize::MAX)
            .await
            .expect("response body should be readable");
        let payload: Value =
            serde_json::from_slice(&body).expect("response body should be valid json");

        assert_eq!(
            payload["analyzed_profile"]["primary_role"],
            json!("mobile_engineer")
        );
        assert_eq!(
            payload["search_profile"]["allowed_sources"],
            json!(["djinni", "work_ua"])
        );
        assert!(
            payload["search_profile"]["target_roles"]
                .as_array()
                .expect("target_roles should be an array")
                .contains(&json!("frontend_engineer"))
        );
    }

    #[tokio::test]
    async fn rejects_invalid_allowed_sources() {
        let result = build_search_profile(
            State(test_state()),
            ApiJson(BuildSearchProfileRequest {
                raw_text: "Senior frontend engineer".to_string(),
                preferences: SearchPreferencesRequest {
                    allowed_sources: vec!["linkedin".to_string()],
                    ..SearchPreferencesRequest::default()
                },
            }),
        )
        .await;

        let response = match result {
            Ok(_) => panic!("handler should reject unknown allowed_sources"),
            Err(error) => error.into_response(),
        };

        assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    }

    #[tokio::test]
    async fn rejects_invalid_preferred_roles() {
        let result = build_search_profile(
            State(test_state()),
            ApiJson(BuildSearchProfileRequest {
                raw_text: "Senior frontend engineer".to_string(),
                preferences: SearchPreferencesRequest {
                    preferred_roles: vec!["frontend_specialist".to_string()],
                    ..SearchPreferencesRequest::default()
                },
            }),
        )
        .await;

        let response = match result {
            Ok(_) => panic!("handler should reject unknown preferred_roles"),
            Err(error) => error.into_response(),
        };

        assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    }
}

File: apps/web/src/api/profiles/search-profile.ts
import { json, request } from '../client';
import type { EngineBuildSearchProfileResponse } from '../engine-types';
import { normalizeMissingString } from '../mappers';
import type { SearchProfileBuildRequest, SearchProfileBuildResult } from './types';
import { DEFAULT_SCORING_WEIGHTS } from './types';

export async function buildSearchProfile(
  payload: SearchProfileBuildRequest,
): Promise<SearchProfileBuildResult> {
	const scoringWeights =
  	payload.preferences?.scoringWeights ?? DEFAULT_SCORING_WEIGHTS;
  const response = await request<EngineBuildSearchProfileResponse>(
    '/api/v1/search-profile/build',
    json('POST', {
      raw_text: payload.rawText,
      preferences: {
			target_regions: payload.preferences?.targetRegions ?? [],
			work_modes: payload.preferences?.workModes ?? [],
			preferred_roles: payload.preferences?.preferredRoles ?? [],
			allowed_sources: payload.preferences?.allowedSources ?? [],
			include_keywords: payload.preferences?.includeKeywords ?? [],
			exclude_keywords: payload.preferences?.excludeKeywords ?? [],
			scoring_weights: {
					skill_match_importance: scoringWeights.skillMatchImportance,
					salary_fit_importance: scoringWeights.salaryFitImportance,
					job_freshness_importance: scoringWeights.jobFreshnessImportance,
					remote_work_importance: scoringWeights.remoteWorkImportance,
				},
			},
    }),
  );

  return {
    analyzedProfile: {
      summary: response.analyzed_profile.summary,
      primaryRole: response.analyzed_profile.primary_role,
      seniority: normalizeMissingString(response.analyzed_profile.seniority) ?? '',
      skills: response.analyzed_profile.skills,
      keywords: response.analyzed_profile.keywords,
      roleCandidates: response.analyzed_profile.role_candidates ?? [],
      suggestedSearchTerms: response.analyzed_profile.suggested_search_terms ?? [],
    },
    searchProfile: {
      primaryRole: response.search_profile.primary_role,
      primaryRoleConfidence: response.search_profile.primary_role_confidence ?? undefined,
      targetRoles: response.search_profile.target_roles,
      roleCandidates: response.search_profile.role_candidates ?? [],
      seniority: normalizeMissingString(response.search_profile.seniority) ?? '',
      targetRegions: response.search_profile.target_regions,
      workModes: response.search_profile.work_modes,
      allowedSources: response.search_profile.allowed_sources,
      profileSkills: response.search_profile.profile_skills ?? [],
      profileKeywords: response.search_profile.profile_keywords ?? [],
      searchTerms: response.search_profile.search_terms,
      excludeTerms: response.search_profile.exclude_terms,
			scoringWeights: {
				skillMatchImportance:
					response.search_profile.scoring_weights?.skill_match_importance ??
					DEFAULT_SCORING_WEIGHTS.skillMatchImportance,
				salaryFitImportance:
					response.search_profile.scoring_weights?.salary_fit_importance ??
					DEFAULT_SCORING_WEIGHTS.salaryFitImportance,
				jobFreshnessImportance:
					response.search_profile.scoring_weights?.job_freshness_importance ??
					DEFAULT_SCORING_WEIGHTS.jobFreshnessImportance,
				remoteWorkImportance:
					response.search_profile.scoring_weights?.remote_work_importance ??
					DEFAULT_SCORING_WEIGHTS.remoteWorkImportance,
			},
    },
  };
}
